### PR TITLE
Use CallerArgumentExpression for internal assert-like constructs

### PR DESCRIPTION
**Pull request by rainersigwald**
[`CallerArgumentExpression`](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-10.0/caller-argument-expression) is a new (C# 10) language feature that allows us to simplify a bunch of our assert-like code to drop the redundant specifications of, like, "which argument is `null`?".

I *strongly* recommend reviewing commit-by-commit; the bulk of the changes are a regex replace on top of a framework that is fairly small.

### CODE REVIEW

File: src/Shared/ErrorUtilities.cs [hunk 1]
```diff
@@ -467,7 +468,7 @@ internal static void VerifyThrowArgument([DoesNotReturnIf(false)] bool condition
         /// Throws an argument out of range exception.
         /// </summary>
         [DoesNotReturn]
-        internal static void ThrowArgumentOutOfRange(string parameterName)
+        internal static void ThrowArgumentOutOfRange(string? parameterName)
```
Comment (JanKrivanek) [inline src/Shared/ErrorUtilities.cs:471]: Why is this needed? - I'd expect this change to not cause weaker nullability guarantee

File: src/Framework/AssemblyUtilities.cs [hunk 2]
```diff
@@ -168,7 +168,7 @@ private static CultureInfo[] GetValidCultures()
 
             var cultures = s_cultureInfoGetCultureMethod.Invoke(null, [allCulturesEnumValue]) as CultureInfo[];
 
-            FrameworkErrorUtilities.VerifyThrowInternalNull(cultures, "CultureInfo.GetCultures should work if all reflection checks pass");
```
Comment (JanKrivanek) [inline src/Framework/AssemblyUtilities.cs:171]: Should we at least keep this as a comment here?

**Review by JanKrivanek — APPROVED** [review]
Great refactoring! Great separation into the reviawable chunks! I love this!

I only have small nonblocking nitpicks

**Review by maridematte — APPROVED** [review]

File: src/Shared/ErrorUtilities.cs (hunk 1 above)
Comment (rainersigwald) [reply src/Shared/ErrorUtilities.cs:471]: Yeah this is a bit unfortunate. To use `[CallerMemberName]` you have to provide a default value, and the one for `string parameterName` that makes the most sense IMO is `null`. But then there's a possibility that `null` flows through the other overload of `ThrowArgumentOutOfRange` to here.

I think I prefer this to swapping it for `""` but if you feel strongly the other way I can swap it. Definitely not a strong preference on my side.

**Comment by rainersigwald** [comment]
```
[mlibc++abi: terminating due to uncaught exception of type PAL_SEHException
Executing [/Users/runner/work/1/s/.dotnet/dotnet "/Users/runner/work/1/s/artifacts/bin/Microsoft.Build.CommandLine.UnitTests/Debug/net9.0/MSBuild.dll" /var/folders/z0/m_s6x1rn5799ttl3khp8n86r0000gn/T/aaev4ycc.fr4/MSBuildTemprunner/Temporary1e6a54b3e9f040aeb48b6b1aa0094d93/testProject.proj]
```

on macOS???

File: src/Shared/ErrorUtilities.cs (hunk 1 above)
Comment (JanKrivanek) [reply src/Shared/ErrorUtilities.cs:471]: Oh - makes sense.
I'm fine with this.
